File: lib/data/isar/models/team.dart
import 'package:isar/isar.dart';

import '../../../core/team_code_generator.dart';

part 'team.g.dart';

@collection
class Team {
  Id id = Isar.autoIncrement;

  @Index(unique: true)
  late String uuid;

  late String name;

  /// 6–8 char uppercase code (no 0,O,1,I). Lookup by this for "Join Team".
  late String inviteCode;

  /// When the owner last rotated the code; re-request after reject is allowed if set and after this time.
  DateTime? inviteCodeRotatedAt;

  /// Coach join code (6–8 chars, uppercase, exclude 0,O,1,I). Owner only can rotate.
  late String coachCode;

  DateTime? coachCodeRotatedAt;

  /// Parent join code (same format). Owner only can rotate.
  late String parentCode;

  DateTime? parentCodeRotatedAt;

  /// Authenticated user ID of the team owner (internal; used for approval).
  String? ownerUserId;

  late DateTime createdAt;

  /// Logo: none | template | monogram | image. Null for teams created before logo fields existed.
  String? logoKind;

  String? templateId;
  String? paletteId;
  String? monogramText;
  /// Reserved for future custom upload.
  String? imagePath;

  /// Server/local last update time (sync).
  late DateTime updatedAt;
  /// User ID of last updater (sync).
  String? updatedBy;
  /// Soft-delete tombstone (optional).
  DateTime? deletedAt;
  /// Schema version for migrations.
  late int schemaVersion;

  Team();

  Team.create({
    required this.uuid,
    required this.name,
    String? inviteCode,
    String? coachCode,
    String? parentCode,
    this.ownerUserId,
    DateTime? createdAt,
    String? logoKind,
    this.templateId,
    this.paletteId,
    this.monogramText,
    this.imagePath,
    DateTime? updatedAt,
    this.updatedBy,
    this.deletedAt,
    int schemaVersion = 1,
  })  : inviteCode = inviteCode ?? TeamCodeGenerator.generate(),
        coachCode = coachCode ?? TeamCodeGenerator.generate(),
        parentCode = parentCode ?? TeamCodeGenerator.generate(),
        createdAt = createdAt ?? DateTime.now(),
        logoKind = logoKind,
        updatedAt = updatedAt ?? DateTime.now(),
        schemaVersion = schemaVersion;
}

File: backend/src/models/Team.js
const mongoose = require('mongoose');
const { v4: uuidv4 } = require('uuid');
const { generateUniqueCode } = require('../utils/codes');

const teamSchema = new mongoose.Schema(
  {
    uuid: { type: String, required: true, unique: true, default: () => uuidv4() },
    name: { type: String, required: true },
    ownerUserId: { type: String, default: null },
    inviteCode: { type: String, required: true },
    inviteCodeRotatedAt: { type: Date, default: null },
    coachCode: { type: String, required: true },
    coachCodeRotatedAt: { type: Date, default: null },
    parentCode: { type: String, required: true },
    parentCodeRotatedAt: { type: Date, default: null },
    createdAt: { type: Date, default: () => new Date() },
    updatedAt: { type: Date, default: () => new Date() },
    updatedBy: { type: String, default: null },
    deletedAt: { type: Date, default: null },
  },
  { timestamps: false }
);

teamSchema.index({ updatedAt: 1 });
teamSchema.index({ deletedAt: 1 });
teamSchema.index({ inviteCode: 1 }, { unique: true });
teamSchema.index({ coachCode: 1 }, { unique: true });
teamSchema.index({ parentCode: 1 }, { unique: true });

teamSchema.statics.createWithCodes = async function (body, userId) {
  const existing = await this.find({}, { inviteCode: 1, coachCode: 1, parentCode: 1 }).lean();
  const codes = new Set(existing.flatMap((t) => [t.inviteCode, t.coachCode, t.parentCode]));
  const now = new Date();
  const normalize = (c) => (typeof c === 'string' && c.trim()) ? c.trim().toUpperCase() : null;
  const inviteCode = normalize(body.inviteCode);
  const coachCode = normalize(body.coachCode);
  const parentCode = normalize(body.parentCode);
  const doc = {
    uuid: body.uuid || uuidv4(),
    name: body.name,
    ownerUserId: userId,
    inviteCode: inviteCode && !codes.has(inviteCode) ? inviteCode : generateUniqueCode(codes),
    coachCode: coachCode && !codes.has(coachCode) ? coachCode : generateUniqueCode(codes),
    parentCode: parentCode && !codes.has(parentCode) ? parentCode : generateUniqueCode(codes),
    createdAt: now,
    updatedAt: now,
    updatedBy: userId,
  };
  const team = await this.create(doc);
  return team;
};

const Team = mongoose.model('Team', teamSchema);
module.exports = { Team };
